lib/Yield/Assembly/DisassemblerInstance.cpp

```diff
@@ -0,0 +1,338 @@
+
+  llvm::StringRef CPUDefinition = "";
+  SubtargetInformation.reset(LLVMTarget->createMCSubtargetInfo(Architecture,
+                                                               CPUDefinition,
+                                                               CPUFeatures));
+  revng_assert(SubtargetInformation != nullptr,
+               "Subtarget information object creation failed.");
+
+  RegisterInformation.reset(LLVMTarget->createMCRegInfo(Architecture));
+  revng_assert(RegisterInformation != nullptr,
+               "Register information object creation failed.");
+
+  llvm::MCTargetOptions TargetOptions;
+  AssemblyInformation.reset(LLVMTarget->createMCAsmInfo(*RegisterInformation,
+                                                        Architecture,
+                                                        TargetOptions));
+  revng_assert(AssemblyInformation != nullptr,
+               "Assembly information object creation failed.");
+
+  ObjectFileInformation = std::make_unique<llvm::MCObjectFileInfo>();
+  Context = std::make_unique<llvm::MCContext>(AssemblyInformation.get(),
+                                              RegisterInformation.get(),
+                                              ObjectFileInformation.get());
+
+  llvm::Triple Triple(Architecture);
+  bool IsPIC = false;
+  ObjectFileInformation->InitMCObjectFileInfo(Triple, IsPIC, *Context);
+
+  auto &SI = *SubtargetInformation;
+  Disassembler.reset(LLVMTarget->createMCDisassembler(SI, *Context));
+  revng_assert(Disassembler != nullptr, "Disassembler object creation failed.");
+
+  InstructionInformation.reset(LLVMTarget->createMCInstrInfo());
+
+  unsigned AssemblyDialect = 0;
+  if (*LLVMArchitecture == llvm::Triple::ArchType::x86
+      || *LLVMArchitecture == llvm::Triple::ArchType::x86_64)
+    if (options::UseIntelSyntax == true)
+      AssemblyDialect = 1;
+  Printer.reset(LLVMTarget->createMCInstPrinter(Triple,
+                                                AssemblyDialect,
+                                                *AssemblyInformation,
+                                                *InstructionInformation,
+                                                *RegisterInformation));
+  revng_assert(Printer != nullptr, "Printer object creation failed.");
+
+  using namespace options;
+  if (ImmediateStyle == ImmediateStyles::Decimal) {
+    Printer->setPrintImmHex(false);
+    Printer->setPrintBranchImmAsAddress(false);
+  } else {
+    Printer->setPrintImmHex(true);
+    Printer->setPrintBranchImmAsAddress(true);
+  }
+
+  if (ImmediateStyle == ImmediateStyles::CHexadecimal)
+    Printer->setPrintHexStyle(llvm::HexStyle::C);
+  else if (ImmediateStyle == ImmediateStyles::AsmHexadecimal)
+    Printer->setPrintHexStyle(llvm::HexStyle::Asm);
+
+  Printer->setSymbolizeOperands(ShouldSymbolizeOperands);
+  Printer->setUseMarkup(true);
+}
+
+static std::pair<std::optional<llvm::MCInst>, uint64_t>
```

Comment:
I'd put the `std::optional` outside.

Reply:
Done: https://github.com/revng/revng/pull/173/commits/99d94861b2bae194d3aafeb42592e101eb6d8902.

Reply:
Actually, I feel like we should go back on this - this can let us properly handle `MCDisassembler::SoftFail`s: the current instruction might be unrecoverable, but we can still proceed recovering the next ones!

What I mean is, the instruction size part of the output (`uint64_t`) can still have meaning even if the disassembly failed.

Any objections?

Reply:
Works!